src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperCompletionService.cs

```diff
@@ -316,5 +318,77 @@ void UpdateCompletions(string tagName, TagHelperDescriptor possibleDescriptor)
                 }
             }
         }
+
+        private static IReadOnlyList<TagHelperDescriptor> FilterFullyQualifiedCompletions(IReadOnlyList<TagHelperDescriptor> possibleChildDescriptors)
+        {
+            // Iterate once through the list to tease apart fully qualified and short name TagHelpers
+            var fullyQualifiedTagHelpers = new List<TagHelperDescriptor>();
+            var shortNameTagHelpers = new HashSet<TagHelperDescriptor>(ShortNameToFullyQualifiedComparer.Instance);
+            for (var i = 0; i < possibleChildDescriptors.Count; i++)
+            {
+                var descriptor = possibleChildDescriptors[i];
+
+                if (descriptor.IsComponentFullyQualifiedNameMatch())
+                {
+                    fullyQualifiedTagHelpers.Add(descriptor);
+                }
+                else
+                {
+                    shortNameTagHelpers.Add(descriptor);
+                }
+            }
+
+            // Re-combine the short named & fully qualified TagHelpers but filter out any fully qualified TagHelpers that have a short
+            // named representation already.
+            var filteredList = new List<TagHelperDescriptor>(shortNameTagHelpers);
+            for (var i = 0; i < fullyQualifiedTagHelpers.Count; i++)
+            {
+                var fullyQualifiedTagHelper = fullyQualifiedTagHelpers[i];
+
+                if (!shortNameTagHelpers.Contains(fullyQualifiedTagHelper))
+                {
+                    // Unimported completion item that isn't represented in a short named form.
+                    filteredList.Add(fullyQualifiedTagHelper);
+                }
+                else
+                {
+                    // There's already a shortname variant of this item, don't include it.
+                }
+            }
+
+            return filteredList;
+        }
+
+        private class ShortNameToFullyQualifiedComparer : IEqualityComparer<TagHelperDescriptor>
+        {
+            public static readonly ShortNameToFullyQualifiedComparer Instance = new ShortNameToFullyQualifiedComparer();
+
+            public bool Equals(TagHelperDescriptor x, TagHelperDescriptor y)
+            {
+                if (x == y)
+                {
+                    return true;
+                }
+
+                if (x == null || y == null)
+                {
+                    return false;
+                }
+
+                if (!string.Equals(x.Name, y.Name, StringComparison.Ordinal))
+                {
+                    return false;
+                }
+
+                if (!string.Equals(x.AssemblyName, y.AssemblyName, StringComparison.Ordinal))
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            public int GetHashCode(TagHelperDescriptor obj) => obj.Name.GetHashCode();
```

Comment:
Not sure how much it matters here, but should HashCode be a combo of both AssemblyName and Name?

Reply:
It could but I figured it wasn't worth the complexity because 99% of the time the Name will be a sufficient separator. For the other 1% the Equals will be called and serve as the true decider

Reply:
> For the other 1% the Equals will be called and serve as the true decider

What about the [taghelper caching apparatus](https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperDescriptorCache.cs). Without the assembly, could we possibly lose taghelpers when they have the same name from multiple assemblies?

Reply:
So we've kind of bastardized the TagHelper hashcode utilization. Typically HashCode's serve as a mechanism to pre-check if objects "may" be equal. If the hashcodes equal then `Equals` is called; if they aren't equal then it assumes there's no chance the two objects rae equivalent. When I say "bastardized" I'm referring  to how we treat TagHelper hash codes for caching purposes where we use the hash code to determine if two TagHelpers are 100% identical which technically should never be true without calling Equals ; but of course Equals can't be serialized 😄.

Anyhow, back to this specific case we're only using this HashCode impl to compare potential Fully qualified/unfully qualified TagHelpers and don't have any serialization requirements at this specific portion of the layer.

Reply:
Makes sense, thanks for the clarification!